File: admin-ui/src/pages/Widgets.js
import React, { useState } from 'react';
import { useQuery, useMutation, useQueryClient } from 'react-query';
import { widgetsAPI } from '../services/api';
import { useConfig } from '../context/ConfigContext';
import { useAuth } from '../context/AuthContext';
import toast from 'react-hot-toast';
import {
  EyeIcon,
  ClipboardDocumentIcon,
  CheckIcon,
} from '@heroicons/react/24/outline';

export default function Widgets() {
  const [copied, setCopied] = useState(false);
  const { config } = useConfig();
  const { user } = useAuth();
  const queryClient = useQueryClient();

  const { data: widgetConfig, isLoading } = useQuery(
    'widgetConfig',
    widgetsAPI.getWidgetConfig
  );

  const updateMutation = useMutation(
    (config) => widgetsAPI.updateWidgetConfig(config),
    {
      onSuccess: () => {
        queryClient.invalidateQueries('widgetConfig');
        toast.success('Widget configuration updated!');
      },
      onError: (error) => {
        toast.error(error.response?.data?.detail || 'Update failed');
      },
    }
  );

  const [formData, setFormData] = useState({
    widget_title: '',
    widget_subtitle: '',
    primary_color: '',
    secondary_color: '',
    welcome_message: '',
    placeholder_text: '',
    is_enabled: true,
  });

  React.useEffect(() => {
    if (widgetConfig) {
      setFormData({
        widget_title: widgetConfig.widget_title || '',
        widget_subtitle: widgetConfig.widget_subtitle || '',
        primary_color: widgetConfig.primary_color || config.primary_color,
        secondary_color: widgetConfig.secondary_color || config.secondary_color,
        welcome_message: widgetConfig.welcome_message || '',
        placeholder_text: widgetConfig.placeholder_text || '',
        is_enabled: widgetConfig.is_enabled ?? true,
      });
    }
  }, [widgetConfig, config]);

  const handleSubmit = (e) => {
    e.preventDefault();
    updateMutation.mutate(formData);
  };

  const handleInputChange = (e) => {
    const { name, value, type, checked } = e.target;
    setFormData(prev => ({
      ...prev,
      [name]: type === 'checkbox' ? checked : value
    }));
  };

  const embedCode = `<script src="${process.env.REACT_APP_API_URL || 'http://localhost:8000'}/api/v1/widgets/${user?.tenantId}/embed.js"></script>`;

  const copyToClipboard = () => {
    navigator.clipboard.writeText(embedCode);
    setCopied(true);
    toast.success('Embed code copied to clipboard!');
    setTimeout(() => setCopied(false), 2000);
  };

  const openPreview = () => {
    const previewUrl = `${process.env.REACT_APP_API_URL || 'http://localhost:8000'}/api/v1/widgets/${user?.tenantId}/preview`;
    window.open(previewUrl, '_blank');
  };

  if (isLoading) {
    return (
      <div className="flex items-center justify-center h-64">
        <div className="animate-spin rounded-full h-12 w-12 border-b-2 border-primary"></div>
      </div>
    );
  }

  return (
    <div className="space-y-6">
      {/* Header */}
      <div>
        <h1 className="text-2xl font-bold text-gray-900">Widget Configuration</h1>
        <p className="mt-2 text-gray-600">
          Customize your embeddable chat widget and get the code to add to your website.
        </p>
      </div>

      <div className="grid grid-cols-1 lg:grid-cols-2 gap-8">
        {/* Configuration Form */}
        <div className="bg-white shadow rounded-lg">
          <div className="px-6 py-4 border-b border-gray-200">
            <h2 className="text-lg font-medium text-gray-900">Widget Settings</h2>
          </div>
          <form onSubmit={handleSubmit} className="p-6 space-y-6">
            <div>
              <label className="block text-sm font-medium text-gray-700 mb-2">
                Widget Title
              </label>
              <input
                type="text"
                name="widget_title"
                value={formData.widget_title}
                onChange={handleInputChange}
                className="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-indigo-500"
                placeholder="e.g., Cleona Assistant"
              />
            </div>

            <div>
              <label className="block text-sm font-medium text-gray-700 mb-2">
                Widget Subtitle
              </label>
              <input
                type="text"
                name="widget_subtitle"
                value={formData.widget_subtitle}
                onChange={handleInputChange}
                className="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-indigo-500"
                placeholder="e.g., How can I help you?"
              />
            </div>

            <div className="grid grid-cols-2 gap-4">
              <div>
                <label className="block text-sm font-medium text-gray-700 mb-2">
                  Primary Color
                </label>
                <div className="flex space-x-2">
                  <input
                    type="color"
                    name="primary_color"
                    value={formData.primary_color}
                    onChange={handleInputChange}
                    className="w-12 h-10 border border-gray-300 rounded cursor-pointer"
                  />
                  <input
                    type="text"
                    name="primary_color"
                    value={formData.primary_color}
                    onChange={handleInputChange}
                    className="flex-1 px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-indigo-500"
                    placeholder="#2E8B57"
                  />
                </div>
              </div>

              <div>
                <label className="block text-sm font-medium text-gray-700 mb-2">
                  Secondary Color
                </label>
                <div className="flex space-x-2">
                  <input
                    type="color"
                    name="secondary_color"
                    value={formData.secondary_color}
                    onChange={handleInputChange}
                    className="w-12 h-10 border border-gray-300 rounded cursor-pointer"
                  />
                  <input
                    type="text"
                    name="secondary_color"
                    value={formData.secondary_color}
                    onChange={handleInputChange}
                    className="flex-1 px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-indigo-500"
                    placeholder="#4682B4"
                  />
                </div>
              </div>
            </div>

            <div>
              <label className="block text-sm font-medium text-gray-700 mb-2">
                Welcome Message
              </label>
              <textarea
                name="welcome_message"
                value={formData.welcome_message}
                onChange={handleInputChange}
                rows={3}
                className="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-indigo-500"
                placeholder="Hello! How can I assist you today?"
              />
            </div>

            <div>
              <label className="block text-sm font-medium text-gray-700 mb-2">
                Input Placeholder
              </label>
              <input
                type="text"
                name="placeholder_text"
                value={formData.placeholder_text}
                onChange={handleInputChange}
                className="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-indigo-500"
                placeholder="Type your message..."
              />
            </div>

            <div className="flex items-center">
              <input
                type="checkbox"
                name="is_enabled"
                checked={formData.is_enabled}
                onChange={handleInputChange}
                className="h-4 w-4 text-indigo-600 focus:ring-indigo-500 border-gray-300 rounded"
              />
              <label className="ml-2 block text-sm text-gray-900">
                Enable widget (users can access the chat)
              </label>
            </div>

            <button
              type="submit"
              disabled={updateMutation.isLoading}
              className="w-full bg-indigo-600 text-white px-4 py-2 rounded-md hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-indigo-500 disabled:opacity-50"
            >
              {updateMutation.isLoading ? 'Saving...' : 'Save Configuration'}
            </button>
          </form>
        </div>

        {/* Preview & Embed Code */}
        <div className="space-y-6">
          {/* Widget Status */}
          <div className="bg-white shadow rounded-lg p-6">
            <h2 className="text-lg font-medium text-gray-900 mb-4">Widget Status</h2>
            <div className="flex items-center justify-between">
              <div>
                <p className="text-sm text-gray-600">Current Status</p>
                <p className={`text-lg font-medium ${formData.is_enabled ? 'text-green-600' : 'text-red-600'}`}>
                  {formData.is_enabled ? 'Active' : 'Disabled'}
                </p>
              </div>
              <div className={`w-3 h-3 rounded-full ${formData.is_enabled ? 'bg-green-400' : 'bg-red-400'}`}></div>
            </div>
            <div className="mt-4 pt-4 border-t border-gray-200">
              <button
                onClick={openPreview}
                className="flex items-center space-x-2 text-indigo-600 hover:text-indigo-500"
              >
                <EyeIcon className="h-5 w-5" />
                <span>Preview Widget</span>
              </button>
            </div>
          </div>

          {/* Embed Code */}
          <div className="bg-white shadow rounded-lg p-6">
            <h2 className="text-lg font-medium text-gray-900 mb-4">Embed Code</h2>
            <p className="text-sm text-gray-600 mb-4">
              Copy this code and paste it into your website's HTML where you want the chat widget to appear.
            </p>
            <div className="relative">
              <pre className="bg-gray-100 p-4 rounded-md text-sm overflow-x-auto border">
                <code>{embedCode}</code>
              </pre>
              <button
                onClick={copyToClipboard}
                className="absolute top-2 right-2 p-2 text-gray-500 hover:text-gray-700"
              >
                {copied ? (
                  <CheckIcon className="h-5 w-5 text-green-500" />
                ) : (
                  <ClipboardDocumentIcon className="h-5 w-5" />
                )}
              </button>
            </div>
            {copied && (
              <p className="text-sm text-green-600 mt-2">Copied to clipboard!</p>
            )}
          </div>

          {/* Widget Preview */}
          <div className="bg-white shadow rounded-lg p-6">
            <h2 className="text-lg font-medium text-gray-900 mb-4">Widget Preview</h2>
            <div className="border border-gray-200 rounded-lg p-4 bg-gray-50">
              <div 
                className="inline-flex items-center px-4 py-2 rounded-lg shadow-sm text-white text-sm"
                style={{ backgroundColor: formData.primary_color }}
              >
                <span className="mr-2">{config.brand_logo}</span>
                {formData.widget_title || 'Chat Assistant'}
              </div>
              <div className="mt-2 text-sm text-gray-600">
                {formData.widget_subtitle || 'How can I help you?'}
              </div>
              <div className="mt-3 p-3 bg-white rounded border text-sm">
                {formData.welcome_message || 'Hello! How can I assist you today?'}
              </div>
              <div className="mt-3">
                <input
                  type="text"
                  placeholder={formData.placeholder_text || 'Type your message...'}
                  className="w-full px-3 py-2 border border-gray-300 rounded text-sm"
                  disabled
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
}

File: api/app/routers/widgets.py
from fastapi import APIRouter, Depends, HTTPException, status, Request
from fastapi.responses import HTMLResponse, PlainTextResponse
from sqlalchemy.orm import Session
from app.database import get_db
from app.auth.dependencies import get_current_tenant_id
from app.models import WidgetConfig
from app.config import config
from app.utils.rate_limit import rate_limit_widget_endpoints
from pydantic import BaseModel
from typing import Optional
import uuid
import logging

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/widgets", tags=["widgets"])

class WidgetConfigRequest(BaseModel):
    widget_title: Optional[str] = None
    widget_subtitle: Optional[str] = None
    primary_color: Optional[str] = None
    secondary_color: Optional[str] = None
    avatar_url: Optional[str] = None
    welcome_message: Optional[str] = None
    placeholder_text: Optional[str] = None
    is_enabled: Optional[bool] = None
    rate_limit_per_hour: Optional[int] = None
    custom_css: Optional[str] = None

class WidgetConfigResponse(BaseModel):
    tenant_id: str
    widget_title: str
    widget_subtitle: str
    primary_color: str
    secondary_color: str
    avatar_url: Optional[str]
    welcome_message: str
    placeholder_text: str
    is_enabled: bool
    rate_limit_per_hour: int
    custom_css: Optional[str]

@router.get("/config", response_model=WidgetConfigResponse)
async def get_widget_config(
    db: Session = Depends(get_db),
    tenant_id: str = Depends(get_current_tenant_id)
):
    """Get widget configuration for authenticated user"""
    try:
        widget_config = db.query(WidgetConfig).filter(
            WidgetConfig.tenant_id == tenant_id
        ).first()
        
        if not widget_config:
            # Create default config
            widget_config = WidgetConfig(
                id=uuid.uuid4(),
                tenant_id=tenant_id,
                widget_title=f"{config.BRAND_NAME} Assistant",
                widget_subtitle="How can I help you?",
                primary_color=config.PRIMARY_COLOR,
                secondary_color=config.SECONDARY_COLOR,
                welcome_message=f"Hello! I'm your {config.BRAND_NAME} assistant. How can I help you today?",
                placeholder_text="Type your message..."
            )
            db.add(widget_config)
            db.commit()
            db.refresh(widget_config)
        
        return WidgetConfigResponse(
            tenant_id=widget_config.tenant_id,
            widget_title=widget_config.widget_title,
            widget_subtitle=widget_config.widget_subtitle,
            primary_color=widget_config.primary_color,
            secondary_color=widget_config.secondary_color,
            avatar_url=widget_config.avatar_url,
            welcome_message=widget_config.welcome_message,
            placeholder_text=widget_config.placeholder_text,
            is_enabled=widget_config.is_enabled,
            rate_limit_per_hour=widget_config.rate_limit_per_hour,
            custom_css=widget_config.custom_css
        )
        
    except Exception as e:
        logger.error(f"Error getting widget config: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to retrieve widget configuration"
        )

@router.put("/config", response_model=WidgetConfigResponse)
async def update_widget_config(
    request: WidgetConfigRequest,
    db: Session = Depends(get_db),
    tenant_id: str = Depends(get_current_tenant_id)
):
    """Update widget configuration"""
    try:
        widget_config = db.query(WidgetConfig).filter(
            WidgetConfig.tenant_id == tenant_id
        ).first()
        
        if not widget_config:
            # Create new config
            widget_config = WidgetConfig(
                id=uuid.uuid4(),
                tenant_id=tenant_id
            )
            db.add(widget_config)
        
        # Update fields
        update_data = request.dict(exclude_unset=True)
        for field, value in update_data.items():
            setattr(widget_config, field, value)
        
        db.commit()
        db.refresh(widget_config)
        
        return WidgetConfigResponse(
            tenant_id=widget_config.tenant_id,
            widget_title=widget_config.widget_title,
            widget_subtitle=widget_config.widget_subtitle,
            primary_color=widget_config.primary_color,
            secondary_color=widget_config.secondary_color,
            avatar_url=widget_config.avatar_url,
            welcome_message=widget_config.welcome_message,
            placeholder_text=widget_config.placeholder_text,
            is_enabled=widget_config.is_enabled,
            rate_limit_per_hour=widget_config.rate_limit_per_hour,
            custom_css=widget_config.custom_css
        )
        
    except Exception as e:
        logger.error(f"Error updating widget config: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to update widget configuration"
        )

@router.get("/{tenant_id}/config", response_model=WidgetConfigResponse)
@rate_limit_widget_endpoints()
async def get_public_widget_config(
    request: Request,
    tenant_id: str,
    db: Session = Depends(get_db)
):
    """Get public widget configuration (no authentication required)"""
    try:
        widget_config = db.query(WidgetConfig).filter(
            WidgetConfig.tenant_id == tenant_id,
            WidgetConfig.is_enabled == True
        ).first()
        
        if not widget_config:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Widget not found or disabled"
            )
        
        return WidgetConfigResponse(
            tenant_id=widget_config.tenant_id,
            widget_title=widget_config.widget_title,
            widget_subtitle=widget_config.widget_subtitle,
            primary_color=widget_config.primary_color,
            secondary_color=widget_config.secondary_color,
            avatar_url=widget_config.avatar_url,
            welcome_message=widget_config.welcome_message,
            placeholder_text=widget_config.placeholder_text,
            is_enabled=widget_config.is_enabled,
            rate_limit_per_hour=widget_config.rate_limit_per_hour,
            custom_css=widget_config.custom_css
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting public widget config: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to retrieve widget configuration"
        )

@router.get("/{tenant_id}/embed.js", response_class=PlainTextResponse)
@rate_limit_widget_endpoints()
async def get_widget_embed_script(
    request: Request,
    tenant_id: str,
    db: Session = Depends(get_db)
):
    """Generate embeddable JavaScript widget"""
    try:
        widget_config = db.query(WidgetConfig).filter(
            WidgetConfig.tenant_id == tenant_id,
            WidgetConfig.is_enabled == True
        ).first()
        
        if not widget_config:
            return "// Widget not found or disabled"
        
        # Helper function to escape JavaScript strings
        def escape_js_string(s):
            if not s:
                return ""
            return s.replace('\\', '\\\\').replace("'", "\\'").replace('"', '\\"').replace('\n', '\\n').replace('\r', '\\r')
        
        # Generate JavaScript embed code
        js_code = f"""
(function() {{
    // {config.BRAND_NAME} Chat Widget
    const TENANT_ID = '{tenant_id}';
    const API_BASE = '{config.API_BASE_URL}';
    const CONFIG = {{
        title: '{escape_js_string(widget_config.widget_title)}',
        subtitle: '{escape_js_string(widget_config.widget_subtitle)}',
        primaryColor: '{escape_js_string(widget_config.primary_color)}',
        secondaryColor: '{escape_js_string(widget_config.secondary_color)}',
        welcomeMessage: '{escape_js_string(widget_config.welcome_message)}',
        placeholder: '{escape_js_string(widget_config.placeholder_text)}',
        avatarUrl: '{escape_js_string(widget_config.avatar_url or "")}'
    }};
    
    let isOpen = false;
    let sessionId = null;
    let messages = [];
    
    // Create widget container
    function createWidget() {{
        const container = document.createElement('div');
        container.id = 'chat-widget-container';
        container.style.cssText = `
            position: fixed;
            bottom: 20px;
            right: 20px;
            z-index: 10000;
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif;
        `;
        
        // Chat toggle button
        const toggleBtn = document.createElement('div');
        toggleBtn.id = 'chat-widget-toggle';
        toggleBtn.style.cssText = `
            width: 60px;
            height: 60px;
            border-radius: 50%;
            background: ${{CONFIG.primaryColor}};
            color: white;
            display: flex;
            align-items: center;
            justify-content: center;
            cursor: pointer;
            box-shadow: 0 4px 12px rgba(0,0,0,0.3);
            font-size: 24px;
            transition: all 0.3s ease;
        `;
        toggleBtn.innerHTML = '💬';
        toggleBtn.onclick = toggleChat;
        
        // Chat window
        const chatWindow = document.createElement('div');
        chatWindow.id = 'chat-widget-window';
        chatWindow.style.cssText = `
            position: absolute;
            bottom: 80px;
            right: 0;
            width: 350px;
            height: 500px;
            background: white;
            border-radius: 12px;
            box-shadow: 0 8px 25px rgba(0,0,0,0.3);
            display: none;
            flex-direction: column;
            overflow: hidden;
        `;
        
        createChatWindow(chatWindow);
        
        container.appendChild(toggleBtn);
        container.appendChild(chatWindow);
        document.body.appendChild(container);
    }}
    
    function createChatWindow(chatWindow) {{
        // Header
        const header = document.createElement('div');
        header.style.cssText = `
            background: ${{CONFIG.primaryColor}};
            color: white;
            padding: 16px;
            display: flex;
            justify-content: space-between;
            align-items: center;
        `;
        
        const headerText = document.createElement('div');
        headerText.innerHTML = `
            <div style="font-weight: 600; font-size: 16px;">${{CONFIG.title}}</div>
            <div style="font-size: 12px; opacity: 0.8;">${{CONFIG.subtitle}}</div>
        `;
        
        const closeBtn = document.createElement('div');
        closeBtn.innerHTML = '✕';
        closeBtn.style.cssText = `
            cursor: pointer;
            font-size: 18px;
            width: 24px;
            height: 24px;
            display: flex;
            align-items: center;
            justify-content: center;
            border-radius: 50%;
        `;
        closeBtn.onclick = toggleChat;
        
        header.appendChild(headerText);
        header.appendChild(closeBtn);
        
        // Messages container
        const messagesContainer = document.createElement('div');
        messagesContainer.id = 'chat-messages';
        messagesContainer.style.cssText = `
            flex: 1;
            overflow-y: auto;
            padding: 16px;
            background: #f8f9fa;
        `;
        
        // Welcome message
        if (CONFIG.welcomeMessage) {{
            const welcomeMsg = createMessageBubble(CONFIG.welcomeMessage, false);
            messagesContainer.appendChild(welcomeMsg);
        }}
        
        // Input container
        const inputContainer = document.createElement('div');
        inputContainer.style.cssText = `
            padding: 16px;
            border-top: 1px solid #e9ecef;
            background: white;
        `;
        
        const inputWrapper = document.createElement('div');
        inputWrapper.style.cssText = `
            display: flex;
            gap: 8px;
            align-items: end;
        `;
        
        const messageInput = document.createElement('textarea');
        messageInput.id = 'chat-input';
        messageInput.placeholder = CONFIG.placeholder;
        messageInput.style.cssText = `
            flex: 1;
            border: 1px solid #ddd;
            border-radius: 20px;
            padding: 12px 16px;
            resize: none;
            max-height: 100px;
            font-family: inherit;
            font-size: 14px;
            outline: none;
        `;
        
        const sendBtn = document.createElement('button');
        sendBtn.innerHTML = '➤';
        sendBtn.style.cssText = `
            width: 40px;
            height: 40px;
            border: none;
            background: ${{CONFIG.primaryColor}};
            color: white;
            border-radius: 50%;
            cursor: pointer;
            font-size: 16px;
            display: flex;
            align-items: center;
            justify-content: center;
        `;
        
        // Event listeners
        sendBtn.onclick = () => sendMessage(messageInput.value);
        messageInput.onkeypress = (e) => {{
            if (e.key === 'Enter' && !e.shiftKey) {{
                e.preventDefault();
                sendMessage(messageInput.value);
            }}
        }};
        
        inputWrapper.appendChild(messageInput);
        inputWrapper.appendChild(sendBtn);
        inputContainer.appendChild(inputWrapper);
        
        chatWindow.appendChild(header);
        chatWindow.appendChild(messagesContainer);
        chatWindow.appendChild(inputContainer);
    }}
    
    function createMessageBubble(content, isUser) {{
        const bubble = document.createElement('div');
        bubble.style.cssText = `
            margin-bottom: 12px;
            display: flex;
            ${{isUser ? 'justify-content: flex-end;' : 'justify-content: flex-start;'}}
        `;
        
        const message = document.createElement('div');
        message.style.cssText = `
            max-width: 80%;
            padding: 12px 16px;
            border-radius: 18px;
            font-size: 14px;
            line-height: 1.4;
            white-space: pre-wrap;
            ${{isUser 
                ? `background: ${{CONFIG.primaryColor}}; color: white; border-bottom-right-radius: 6px;`
                : 'background: white; color: #333; border: 1px solid #e9ecef; border-bottom-left-radius: 6px;'
            }}
        `;
        message.textContent = content;
        
        bubble.appendChild(message);
        return bubble;
    }}
    
    async function sendMessage(content) {{
        if (!content.trim()) return;
        
        const messageInput = document.getElementById('chat-input');
        const messagesContainer = document.getElementById('chat-messages');
        
        // Clear input
        messageInput.value = '';
        
        // Add user message
        const userBubble = createMessageBubble(content, true);
        messagesContainer.appendChild(userBubble);
        messagesContainer.scrollTop = messagesContainer.scrollHeight;
        
        // Add loading indicator
        const loadingBubble = createMessageBubble('...', false);
        loadingBubble.id = 'loading-message';
        messagesContainer.appendChild(loadingBubble);
        messagesContainer.scrollTop = messagesContainer.scrollHeight;
        
        try {{
            const response = await fetch(`${{API_BASE}}/api/v1/chat/${{TENANT_ID}}/query`, {{
                method: 'POST',
                headers: {{
                    'Content-Type': 'application/json',
                }},
                body: JSON.stringify({{
                    message: content,
                    session_id: sessionId
                }})
            }});
            
            if (!response.ok) {{
                throw new Error('Failed to send message');
            }}
            
            const data = await response.json();
            
            // Remove loading message
            const loading = document.getElementById('loading-message');
            if (loading) loading.remove();
            
            // Add assistant response
            const assistantBubble = createMessageBubble(data.answer, false);
            messagesContainer.appendChild(assistantBubble);
            messagesContainer.scrollTop = messagesContainer.scrollHeight;
            
            // Store session ID
            if (data.session_id && !sessionId) {{
                sessionId = data.session_id;
            }}
            
        }} catch (error) {{
            console.error('Chat error:', error);
            
            // Remove loading message
            const loading = document.getElementById('loading-message');
            if (loading) loading.remove();
            
            // Show error message
            const errorBubble = createMessageBubble('Sorry, I encountered an error. Please try again.', false);
            messagesContainer.appendChild(errorBubble);
            messagesContainer.scrollTop = messagesContainer.scrollHeight;
        }}
    }}
    
    function toggleChat() {{
        const chatWindow = document.getElementById('chat-widget-window');
        const toggleBtn = document.getElementById('chat-widget-toggle');
        
        isOpen = !isOpen;
        
        if (isOpen) {{
            chatWindow.style.display = 'flex';
            toggleBtn.innerHTML = '✕';
        }} else {{
            chatWindow.style.display = 'none';
            toggleBtn.innerHTML = '💬';
        }}
    }}
    
    // Initialize widget when DOM is ready
    if (document.readyState === 'loading') {{
        document.addEventListener('DOMContentLoaded', createWidget);
    }} else {{
        createWidget();
    }}
}})();
"""
        
        return js_code
        
    except Exception as e:
        logger.error(f"Error generating widget embed script: {e}")
        return "// Error loading widget"

@router.get("/{tenant_id}/preview", response_class=HTMLResponse)
@rate_limit_widget_endpoints()
async def get_widget_preview(
    request: Request,
    tenant_id: str,
    db: Session = Depends(get_db)
):
    """Generate HTML preview of the widget"""
    try:
        widget_config = db.query(WidgetConfig).filter(
            WidgetConfig.tenant_id == tenant_id
        ).first()
        
        if not widget_config:
            return "<html><body><h1>Widget not found</h1></body></html>"
        
        html = f"""
<!DOCTYPE html>
<html>
<head>
    <title>{widget_config.widget_title} - Preview</title>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <style>
        body {{
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif;
            margin: 0;
            padding: 20px;
            background-color: #f5f5f5;
        }}
        .preview-container {{
            max-width: 800px;
            margin: 0 auto;
            background: white;
            padding: 20px;
            border-radius: 8px;
            box-shadow: 0 2px 10px rgba(0,0,0,0.1);
        }}
        .widget-info {{
            background: {widget_config.primary_color};
            color: white;
            padding: 15px;
            border-radius: 5px;
            margin-bottom: 20px;
        }}
        .embed-code {{
            background: #f8f9fa;
            padding: 15px;
            border-radius: 5px;
            border: 1px solid #e9ecef;
            font-family: monospace;
            white-space: pre-wrap;
            word-break: break-all;
        }}
    </style>
</head>
<body>
    <div class="preview-container">
        <h1>{config.BRAND_NAME} Chat Widget Preview</h1>
        
        <div class="widget-info">
            <h2>{widget_config.widget_title}</h2>
            <p>{widget_config.widget_subtitle}</p>
            <p><strong>Welcome Message:</strong> {widget_config.welcome_message}</p>
            <p><strong>Status:</strong> {'Enabled' if widget_config.is_enabled else 'Disabled'}</p>
        </div>
        
        <h3>Embed Code</h3>
        <p>Copy and paste this code into your website:</p>
        <div class="embed-code">&lt;script src="{config.API_BASE_URL}/api/v1/widgets/{tenant_id}/embed.js"&gt;&lt;/script&gt;</div>
        
        <h3>API Endpoints</h3>
        <ul>
            <li><strong>Chat:</strong> POST /api/v1/chat/{tenant_id}/query</li>
            <li><strong>Config:</strong> GET /api/v1/widgets/{tenant_id}/config</li>
        </ul>
    </div>
    
    <!-- Load the actual widget -->
    <script src="{config.API_BASE_URL}/api/v1/widgets/{tenant_id}/embed.js"></script>
</body>
</html>
"""
        
        return html
        
    except Exception as e:
        logger.error(f"Error generating widget preview: {e}")
        return "<html><body><h1>Error loading preview</h1></body></html>"